File: 1103.distribute-candies-to-people.cpp
/*
 * @lc app=leetcode id=1103 lang=cpp
 *
 * [1103] Distribute Candies to People
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  vector<int> distributeCandies(int candies, int num_people) {
    vector<int> ret(num_people, 0);
    int count = 1;
    while (candies > 0) {
      for (int i = 0; i < num_people; i++) {
        int increment = std::min(count, candies);
        ret[i] += increment;
        candies -= increment;
        count++;
      }
    }
    return ret;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<int> exp;
  exp = {1, 2, 3, 1};
  ASSERT_EQ(s.distributeCandies(7, 4), exp);
  exp = {5, 2, 3};
  ASSERT_EQ(s.distributeCandies(10, 3), exp);
  return 0;
}

File: 1103.distribute-candies-to-people.rs
/*
 * @lc app=leetcode id=1103 lang=rust
 *
 * [1103] Distribute Candies to People
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn distribute_candies(mut candies: i32, num_people: i32) -> Vec<i32> {
        let mut ret = Vec::<i32>::new();
        ret.resize(num_people as usize, 0);
        let mut i = 0;
        while candies > 0 {
            ret[(i % num_people) as usize] += (i + 1).min(candies);
            i += 1;
            candies -= i;
        }
        ret
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::distribute_candies(7, 4), vec![1, 2, 3, 1]);
}

#[test]
fn test2() {
    assert_eq!(Solution::distribute_candies(10, 3), vec![5, 2, 3]);
}
